### ISSUE TITLE
Design: Improve MD Report

### ISSUE THREAD

**Issue by BioPhoton** [flowup/quality-metrics-cli#93]
**Todos:**

- linebreaks:
  - line breaks before each heading 
  - add line break after the report
- hash correct in data?
- overview table header bold
- only reports with score 1 pass
- score is 0 - 100
- leaf out optional fields if `undefined`
- category detail section
  - Dummy Category 1 **- 7** => the dash score is not clear
- audit detail => score unused

**Comment by matejchalk** [comment]
@BioPhoton I've put together a small example of how the Markdown report could be formatted. What do you think?


---

# Code PushUp report

| Category                    | Score  | Audits |
| :-------------------------- | :----: | :----: |
| [Performance](#performance) | **92** |   5    |
| [Code style](#code-style)   | **54** |   12   |

| Commit                                                                                                                  | Version | Duration | Plugins | Categories | Audits |
| :---------------------------------------------------------------------------------------------------------------------- | :------ | :------: | :-----: | :--------: | :----: |
| _Implement todos list_ ([af75415](https://github.com/flowup/todos-app/commit/af75415ec34a16e816b611231f24a72e7f5fc781)) | `0.1.0` |  1.6 s   |    2    |     3      |   25   |

| Plugin                                                                                          | Audits | Duration |
| :---------------------------------------------------------------------------------------------- | :----: | :------: |
| [ESLint](https://github.com/flowup/quality-metrics-cli/tree/main/packages/plugin-eslint#readme) |   20   |  0.3 s   |
| Lighthouse                                                                                      |   5    |  1.2 s   |

## Categories

### Performance

Score: **92**

- Performance (_Lighthouse_):
  - [Largest Contentful Paint](#largest-contentful-paint-lighthouse) - **1.5 s**
- [Disallow missing `key` props in iterators/collection literals](#disallow-missing-key-props-in-iteratorscollection-literals-eslint) (_ESLint_) - **1 error**

### Code style

Score: **54**

- [Require or disallow method and property shorthand syntax for object literals](#require-or-disallow-method-and-property-shorthand-syntax-for-object-literals-eslint) (_ESLint_) - **4 warnings**
- [Require `let` or `const` instead of `var`](#require-let-or-const-instead-of-var-eslint) (_ESLint_) - **0**

## Audits

### Largest Contentful Paint (Lighthouse)

[Link to docs](https://developer.chrome.com/docs/lighthouse/performance/lighthouse-largest-contentful-paint/)

<b>1.5 s</b> (score: 75)

### Disallow missing `key` props in iterators/collection literals (ESLint)

ESLint rule **jsx-key**, from _react_ plugin.

[Link to docs](https://github.com/jsx-eslint/eslint-plugin-react/tree/master/docs/rules/jsx-key.md)

<details>
  <summary><b>1 error</b> (score: 0)</summary>
  <h4>Issues</h4>
  <table>
    <tr>
      <th>Severity</th>
      <th>Message</th>
      <th>Source file</th>
      <th>Line(s)</th>  
    </tr>
    <tr>
      <td><i>error</i></td>
      <td>Missing "key" prop for element in iterator</td>
      <td><code>src/components/TodoList.jsx</code></td>
      <td>7</td>
    </tr>
  </table>
</details>

### Require or disallow method and property shorthand syntax for object literals (ESLint)

ESLint rule **object-shorthand**.

[Link to docs](https://eslint.org/docs/latest/rules/object-shorthand)

<details>
  <summary><b>4 warnings</b> (score: 0)</summary>
  <h4>Issues</h4>
  <table>
    <tr>
      <th>Severity</th>
      <th>Message</th>
      <th>Source file</th>
      <th>Line(s)</th>  
    </tr>
    <tr>
      <td><i>warning</i></td>
      <td>Expected property shorthand</td>
      <td><code>src/hooks/useTodos.js</code></td>
      <td>19</td>
    </tr>
    <tr>
      <td><i>warning</i></td>
      <td>Expected property shorthand</td>
      <td><code>src/hooks/useTodos.js</code></td>
      <td>32</td>
    </tr>
    <tr>
      <td><i>warning</i></td>
      <td>Expected property shorthand</td>
      <td><code>src/hooks/useTodos.js</code></td>
      <td>33</td>
    </tr>
  </table>
</details>

### Require `let` or `const` instead of `var` (ESLint)

ESLint rule **no-var**.

[Link to docs](https://eslint.org/docs/latest/rules/no-var)

<b>0</b> (score: 100)

**Comment by BioPhoton** [comment]
is Source file linkable?

**Comment by BioPhoton** [comment]
Plugin | Audits | Duration
-- | -- | --


this info could be collapsed with other more detailed infos to keep the report small?

**Comment by BioPhoton** [comment]
<h4 dir="auto" style="box-sizing: border-box; margin-top: 24px; margin-bottom: 16px; font-size: 14px; font-weight: var(--base-text-weight-semibold, 600); line-height: 1.25; color: rgb(31, 35, 40); font-family: -apple-system, &quot;system-ui&quot;, &quot;Segoe UI&quot;, &quot;Noto Sans&quot;, Helvetica, Arial, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; white-space: normal; text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">Issues</h4>

Severity | Message | Source file | Line(s)
-- | -- | -- | --



The Issue table will not scale very well when we have longer descriptions from lighthouse or the nx-validators

**Comment by BioPhoton** [comment]
Also we should see how the vertical space grows when we have a full report

**Comment by matejchalk** [comment]
> Is source file in the issues linkeable to the line of code?

It could be, yes. Similar as for the commit, it all depends on if we can parse their remote origin from Git (GitHub/Gitlab/... - could be tricky for self-hosted repos). Then we can link to both the commit and file lines in that commit.

>  Plugin | Audits | Duration
> -- | -- | --
> 
> this info could be collapsed with other more detailed infos to keep the report small?

Could be. I think the 1st table with category scores should be the most prominent. The next two tables are some other general info, which is good to have somewhere, but usually not the most interesting part. We could put it at the end, I suppose (most probably won't scroll that far) :thinking: 

> The Issue table will not scale very well when we have longer descriptions from lighthouse or the nx-validators

Depends how long the message texts are. For lint errors, they're usually one short sentence. The text in cells is wrapped automatically, so I think it'll look OK even if there are 2 or 3 lines used. We could also have a character limit (replace overflowed text with `...`) if it turns out some messages are really long.

> Also we should see how the vertical space grows when we have a full report

Yes, the scaling will be interesting. I'd be a little concerned with the amount of audits (having hundreds of ESLint rules is fairly realistic, for example), and most concerned with the issues (some rules will have hundreds or thousands of errors in large codebases). The issues are always collapsed, so visually that shouldn't be an issue. But the file size could get out of hand. A possible solution would be to generate multiple Markdown files (e.g. one for each audit), but lets cross that bridge when we get there (might also be worth considering HTML output at that point :sweat_smile:).

**Comment by matejchalk** [comment]
TBD:

# Code PushUp report

| Category                    | Score  | Audits |
| :-------------------------- | :----: | :----: |
| [Performance](#performance) | **92** |   5    |
| [Code style](#code-style)   | **54** |   12   |

## Categories

### Performance

Score: **92**

- Performance (_Lighthouse_):
  - ❌ [Largest Contentful Paint](#largest-contentful-paint-lighthouse) - **1.5 s**
-  ❌ [Disallow missing `key` props in iterators/collection literals](#disallow-missing-key-props-in-iteratorscollection-literals-eslint) (_ESLint_) - **1 error**

### Code style

Score: **54**

- ❌ [Require or disallow method and property shorthand syntax for object literals](#require-or-disallow-method-and-property-shorthand-syntax-for-object-literals-eslint) (_ESLint_) - **4 warnings**
- ✅ [Require `let` or `const` instead of `var`](#require-let-or-const-instead-of-var-eslint) (_ESLint_) - **0**

## Audits

### Largest Contentful Paint (Lighthouse)

[📖 Link to docs](https://developer.chrome.com/docs/lighthouse/performance/lighthouse-largest-contentful-paint/)

<b>1.5 s</b> (score: 75)

### Disallow missing `key` props in iterators/collection literals (ESLint)

ESLint rule **jsx-key**, from _react_ plugin.

[📖 Link to docs](https://github.com/jsx-eslint/eslint-plugin-react/tree/master/docs/rules/jsx-key.md)

<details>
  <summary><b>1 error</b> (score: 0)</summary>
  <h4>Issues</h4>
  <table>
    <tr>
      <th>Severity</th>
      <th>Message</th>
      <th>Source file</th>
      <th>Line(s)</th>  
    </tr>
    <tr>
      <td><i>error</i></td>
      <td>Missing "key" prop for element in iterator</td>
      <td><code>src/components/TodoList.jsx</code></td>
      <td>7</td>
    </tr>
  </table>
</details>

### Require or disallow method and property shorthand syntax for object literals (ESLint)

ESLint rule **object-shorthand**.

[📖 Link to docs](https://eslint.org/docs/latest/rules/object-shorthand)

<details>
  <summary><b>4 warnings</b> (score: 0)</summary>
  <h4>Issues</h4>
  <table>
    <tr>
      <th>Severity</th>
      <th>Message</th>
      <th>Source file</th>
      <th>Line(s)</th>  
    </tr>
    <tr>
      <td><i>warning</i></td>
      <td>Expected property shorthand</td>
      <td><code>src/hooks/useTodos.js</code></td>
      <td>19</td>
    </tr>
    <tr>
      <td><i>warning</i></td>
      <td>Expected property shorthand</td>
      <td><code>src/hooks/useTodos.js</code></td>
      <td>32</td>
    </tr>
    <tr>
      <td><i>warning</i></td>
      <td>Expected property shorthand</td>
      <td><code>src/hooks/useTodos.js</code></td>
      <td>33</td>
    </tr>
  </table>
</details>

### Require `let` or `const` instead of `var` (ESLint)

ESLint rule **no-var**.

[📖 Link to docs](https://eslint.org/docs/latest/rules/no-var)

<b>0</b> (score: 100)

## Metadata

| Commit                                                                                                                  | Version | Duration | Plugins | Categories | Audits |
| :---------------------------------------------------------------------------------------------------------------------- | :------ | :------: | :-----: | :--------: | :----: |
| _Implement todos list_ ([af75415](https://github.com/flowup/todos-app/commit/af75415ec34a16e816b611231f24a72e7f5fc781)) | `0.1.0` |  1.6 s   |    2    |     3      |   25   |

| Plugin                                                                                          | Audits | Duration |
| :---------------------------------------------------------------------------------------------- | :----: | :------: |
| [ESLint](https://github.com/flowup/quality-metrics-cli/tree/main/packages/plugin-eslint#readme) |   20   |  0.3 s   |
| Lighthouse                                                                                      |   5    |  1.2 s   |

**Comment by matejchalk** [comment]
@BioPhoton @vmasek 

I created a separate issue for the implementation that we can refine: #148
I separated the git web URLs, since they're quite complex to implement and not a must have, in my opinion: #149

To be clear, the goal for me is for a CLI user to get "something meaningful" without having to connect to the portal. The `report.json` is too raw, but `report.md` should be readable. It's still gonna "look like Markdown" though (obviously), if we want something really stylized then that we should pick a different format (`report.html` or even `report.pdf`), but I think that's overkill for now.

The main problem I want to address in that I don't think we can confidently ship the current `report.md` (screenshot below), but isn't logically structured, misses crucial information (e.g. audit results) and has broken formatting in general.

![Image](https://github.com/flowup/quality-metrics-cli/assets/34691111/e57aa48b-42fc-4827-8bbb-307f231b2218)

As far posting it as a comment in PR, we could support an abbreviated version, I guess (only title and category scores table). But honestly, I don't think that workflow is well thought out. PR comments are a very different feature for me, because they're only valuable if they show you what changed.
